overlord/snapstate/snapstate.go

```diff
@@ -4146,6 +4147,125 @@ func GadgetConnections(st *state.State, deviceCtx DeviceContext) ([]gadget.Conne
 	return gadgetInfo.Connections, nil
 }
 
+// deriveNeededSnapRevisions returns revisions that need to be kept for the given snap.
+//
+// A revision is needed only if:
+//    1. Revision is in a snap's sequence
+//    2. Revision is in a snap's refresh hint
+//    3. Revision is pointed to by a download task in an ongoing change
```

Comment:
nitpick: I'd add just some references so that someone can go look at those places to see what purpose they serve
```suggestion
//    1. The revision is in SnapState.Sequence
//    2. The revision is in saved in refresh-candidates
//    3. The revision is pointed to by a download task in an ongoing change
```